src/capmap.cc

```diff
@@ -0,0 +1,156 @@
+// SPDX-FileCopyrightText: Copyright 2023 Arm Limited and/or its affiliates <open-source-office@arm.com>
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#include "include/capmap.h"
+
+#include <inttypes.h>
+#include <stdio.h>
+
+// BSD headers for kinfo_getvmmap.
+#include <libutil.h>
+#include <sys/types.h>
+#include <sys/user.h>
+#include <unistd.h>
+
+#if __has_feature(capabilities)
+#include <cheriintrin.h>
+#else
+#error "capmap.cc requires capabilities"
+#endif
+
+namespace capmap {
+
+void print_raw_cap(FILE* stream, void* __capability cap) {
+  uint64_t parts[2];
+  static_assert(sizeof(parts) == sizeof(cap), "Expected 128-bit capability");
+  memcpy(parts, &cap, sizeof(parts));
+  fprintf(stream, "0x%u:%" PRIx64 ":%" PRIx64, cheri_tag_get(cap), parts[1], parts[0]);
+}
+
+void Mapper::update_self_ranges() {
+  // TODO: Do a better job here.
+  //  - We have several heap objects, which should also be excluded.
+  //  - If using scan(Roots), we might want to exclude invariant regions once,
+  //    and update the exclusion ranges only when we allocate new map entries,
+  //    etc.
+  exclude_self_ = SparseRange();
+  exclude_self_.combine(Range::from_object(this));
+}
+
+void Mapper::scan(void* __capability cap, uint64_t depth) {
+  // fprintf(stderr, "scan(%#lp, %" PRIu64 ")\n", cap, depth);
+  if (depth > max_seen_scan_depth_) max_seen_scan_depth_ = depth;
```

Comment:
`max_scan_depth_` is checked when we're about to scan for recursion opportunities (line 59). Checking there avoids a lot of needless work if we hit the limit. However, we bump `max_seen_scan_depth_` only if we actually do recurse.

Reply:
> line 59

That for me (in `capmap.cc`) looks like an assembly instruction. But if you know that there is a check done somewhere in a sub-call, then it's fine.

Reply:
Perhaps things moved when I updated it. I meant this line:

```
if (load_cap_map_.try_combine(cap) && (depth < max_scan_depth_)) {
```

* If `cap` isn't already in our map (so we ignore loops) then add it.
* If we also haven't already hit `max_scan_depth_`, then scan it.

The order of operations and checks is important, hence all the tests and debug code :-) We have to `try_combine()` before we check the depth because we still want to record `cap` itself, even if we aren't going to recurse. We could put the depth check at the top of `scan()` (so it happens at a lower level) but then we'd go through the trouble of loading all the a region's contained capabilities, but ignore all of them.

I'm sure this could be better optimised — I didn't apply any real rigour — but that can happen later, if necessary.